File: src/main/java/net/atomique/ksar/UI/SortedTreeNode.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package net.atomique.ksar.UI;

import java.util.Collections;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.MutableTreeNode;

/**
 *
 * @author Max
 */
public class SortedTreeNode extends DefaultMutableTreeNode implements Comparable {

    public static final long serialVersionUID = 15071L;


    public SortedTreeNode(String name) {
        super(name);
    }

    public SortedTreeNode(TreeNodeInfo tmp) {
        super(tmp);
    }

    public SortedTreeNode(ParentNodeInfo tmp) {
        super(tmp);
    }
    /*
    public SortedTreeNode (GraphDescription graphdesc) {
        super(graphdesc);
    }
    */

    public void count_graph(SortedTreeNode node) {
        int num = node.getChildCount();

        if (num > 0) {
            for (int i = 0; i < num; i++) {
                SortedTreeNode l = (SortedTreeNode) node.getChildAt(i);
                count_graph(l);
            }
        } else {
            Object obj1 = node.getUserObject();
            if (obj1 instanceof TreeNodeInfo) {
                leaf_num++;
            }
        }
        
    }
    
    public int  LeafCount() {
        leaf_num=0;
        count_graph(this);
        return leaf_num;
    }
    
    @Override
    public void insert(final MutableTreeNode newChild, final int childIndex) {
        super.insert(newChild, childIndex);
        Collections.sort(this.children);
    }
    public int compareTo(final Object o) {
        return this.toString().compareToIgnoreCase(o.toString());
    }

    private int leaf_num=0;
}

File: src/main/kotlin/net/atomique/ksar/ui/SortedTreeNode.kt
/*
 * Copyright 2018 The kSAR Project. All rights reserved.
 * See the LICENSE file in the project root for more information.
 */
package net.atomique.ksar.ui

import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.MutableTreeNode
import java.util.*

class SortedTreeNode : DefaultMutableTreeNode, Comparable<Any?> {
    private var leaf_num = 0

    companion object {
        const val serialVersionUID = 15071L
        private val comparator = Comparator.nullsFirst(
            Comparator.comparing { v: String -> "all" == v }
                .thenComparing { v -> "sum" == v }
                .reversed()
                .thenComparing(NaturalComparator)
        )
    }

    constructor(name: String?) : super(name) {}
    constructor(tmp: TreeNodeInfo?) : super(tmp) {}
    constructor(tmp: ParentNodeInfo?) : super(tmp) {}

    /*
  public SortedTreeNode (GraphDescription graphdesc) {
      super(graphdesc);
  }
  */
    fun count_graph(node: SortedTreeNode) {
        val num = node.childCount
        if (num > 0) {
            for (i in 0 until num) {
                val l = node.getChildAt(i) as SortedTreeNode
                count_graph(l)
            }
        } else {
            val obj1 = node.getUserObject()
            if (obj1 is TreeNodeInfo) {
                leaf_num++
            }
        }
    }

    fun LeafCount(): Int {
        leaf_num = 0
        count_graph(this)
        return leaf_num
    }

    override fun insert(newChild: MutableTreeNode, childIndex: Int) {
        super.insert(newChild, childIndex)
        (children as Vector<SortedTreeNode>).sort()
    }

    override fun compareTo(other: Any?): Int {
        return comparator.compare(this.toString(), other.toString())
    }
}
